File: src/motionlab_gait/analysis/normal_reference.py
from __future__ import annotations

from collections.abc import Sequence

import numpy as np

REFERENCE_LABEL = "成人・平地・快適歩行の概略参考（診断基準ではありません）"
REFERENCE_SOURCES = (
    "Michelini et al. Prosthet Orthot Int. 2020;44:245-262. PMID: 32507049",
    "Stenum et al. PLoS Comput Biol. 2021;17:e1008935. PMID: 33891585",
    "成人歩行の概略ROM: 股関節約-10〜35°、膝約5〜60°、足関節約-15〜15°",
)

# Broad illustrative anchors for level, comfortable adult walking. These are intentionally
# wider than a single laboratory's mean curve because 2D MediaPipe angles are not identical
# to a calibrated 3D biomechanical model.
_ANCHORS: dict[str, tuple[Sequence[float], Sequence[float], float]] = {
    "hip_flexion": (
        (0, 10, 30, 50, 60, 75, 100),
        (28, 24, 8, -8, -10, 12, 28),
        10.0,
    ),
    "knee_flexion": (
        (0, 10, 30, 50, 60, 75, 100),
        (5, 15, 5, 10, 35, 60, 5),
        12.0,
    ),
    "ankle_dorsiflexion": (
        (0, 10, 30, 50, 60, 70, 85, 100),
        (0, -7, 5, 10, -15, -5, 3, 0),
        8.0,
    ),
}


def normal_reference(joint: str) -> tuple[tuple[float, float, float, float], ...]:
    percentages, means, half_width = _ANCHORS[joint]
    x = np.arange(101, dtype=float)
    mean = np.interp(x, percentages, means)
    return tuple(
        (float(percent), float(value), float(value - half_width), float(value + half_width))
        for percent, value in zip(x, mean, strict=True)
    )

File: web/src/analysis/normalReference.ts
import type { Joint } from "./types";

export const REFERENCE_LABEL = "成人・平地・快適歩行の概略参考（診断基準ではありません）";

const ANCHORS: Record<Joint, { percentages: number[]; means: number[]; halfWidth: number }> = {
  hip_flexion: { percentages: [0, 10, 30, 50, 60, 75, 100], means: [28, 24, 8, -8, -10, 12, 28], halfWidth: 10 },
  knee_flexion: { percentages: [0, 10, 30, 50, 60, 75, 100], means: [5, 15, 5, 10, 35, 60, 5], halfWidth: 12 },
  ankle_dorsiflexion: { percentages: [0, 10, 30, 50, 60, 70, 85, 100], means: [0, -7, 5, 10, -15, -5, 3, 0], halfWidth: 8 }
};

export function normalReference(joint: Joint): { percent: number; mean: number; lower: number; upper: number }[] {
  const source = ANCHORS[joint];
  return Array.from({ length: 101 }, (_, percent) => {
    let right = source.percentages.findIndex((value) => value >= percent);
    if (right < 0) right = source.percentages.length - 1;
    const left = Math.max(0, right - 1);
    const x0 = source.percentages[left]!; const x1 = source.percentages[right]!;
    const y0 = source.means[left]!; const y1 = source.means[right]!;
    const value = x1 === x0 ? y0 : y0 + ((percent - x0) / (x1 - x0)) * (y1 - y0);
    return { percent, mean: value, lower: value - source.halfWidth, upper: value + source.halfWidth };
  });
}

